### PR TITLE
Feature/delay intermediate commits during multi operations

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Do not execute intermediate commits in the middle of array/multi-document operations.
This change will delay any intermediate commit until the end of the array operation. This is necessary so that the synchronous replication can be carried out while the transaction's document keys are still fully locked on the leader and dictate the sequence of replicated events for any given key.

- [x] :hankey: Bugfix (requires CHANGELOG entry)
- [x] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [x] Backport for 3.9: TO DO

### Testing & Verification

- [x] The behavior in this PR was *manually tested*
- [x] This change is already covered by existing tests, such as *shell_client*, *shell_server*.

### CODE REVIEW

File: arangod/Transaction/Methods.cpp [hunk 1]
```diff
@@ -2827,6 +2861,12 @@ futures::Future<OperationResult> Methods::countInternal(std::string const& colle
   return futures::makeFuture(countLocal(collectionName, type, options));
 }
 
+// perform a (deferred) intermediate commit if required
+Result Methods::performIntermediateCommitIfRequired(DataSourceId collectionId) {
+  [[maybe_unused]] bool hasPerformedIntermediateCommit = false;
+  return _state->performIntermediateCommitIfRequired(collectionId, hasPerformedIntermediateCommit);
```
Comment (mpoeter) [inline arangod/Transaction/Methods.cpp:2867]: I think I found no caller that actually uses the value stored in `hasPerformedIntermediateCommit`. Can we get rid for that parameter?

File: tests/js/client/shell/shell-transaction-intermediate-commit-babies-cluster.js [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+      let didWork = false;
+      try {
+        db._query('FOR i IN 1..10000 INSERT { _key: CONCAT("test", i), value: ASSERT(i < 10000, "peng!") } IN ' + cn, {}, {intermediateCommitCount: 100});
+        didWork = true;
+      } catch (err) {
+      }
+      assertFalse(didWork);
+      
+      let droppedFollowersAfter = getMetric(leader, "arangodb_dropped_followers_total");
+      assertEqual(droppedFollowersBefore + 1, droppedFollowersAfter);
```
Comment (mpoeter) [inline tests/js/client/shell/shell-transaction-intermediate-commit-babies-cluster.js:179]: Why do we expect a dropped follower in this case?

File: tests/js/client/shell/shell-transaction-intermediate-commit-babies-cluster.js [hunk 3]
```diff
@@ -0,0 +1,306 @@
...
+      assertEqual(intermediateCommitsBefore, intermediateCommitsAfter);
+        
+      assertEqual(1000, db._collection(cn).count());
+    },
+    
+    testDocumentsTransactionIntermediateCommitsSuppressed: function () {
+      let c = db._create(cn, { numberOfShards: 1, replicationFactor: 1 });
+      let shards = db._collection(cn).shards(true);
+      let shardId = Object.keys(shards)[0];
+      let leader = getEndpointById(shards[shardId][0]);
```
Comment (mpoeter) [inline tests/js/client/shell/shell-transaction-intermediate-commit-babies-cluster.js:272]: This initialization code is almost identical for all testcases, though some use rf=2 and some rf=1. I would prefer if we could split these tests into two separate testsuites and move the common initialization code into the respective setup functions.

**Review by mpoeter — APPROVED** [review]

File: tests/js/client/shell/shell-transaction-intermediate-commit-babies-cluster.js (hunk 2 above)
Comment (jsteemann) [reply tests/js/client/shell/shell-transaction-intermediate-commit-babies-cluster.js]: Added a comment to the PR.

File: tests/js/client/shell/shell-transaction-intermediate-commit-babies-cluster.js (hunk 3 above)
Comment (jsteemann) [reply tests/js/client/shell/shell-transaction-intermediate-commit-babies-cluster.js]: Actually the setups are slightly different. I thus created 2 separate testsuites out of it so it becomes a bit clearer.

File: arangod/Transaction/Methods.cpp (hunk 1 above)
Comment (jsteemann) [reply arangod/Transaction/Methods.cpp]: Good catch. Fixed!

**Review by mpoeter — APPROVED** [review]

**Pull request merged**